### PR TITLE
[onert] Load tensor and optimizer data from the checkpoint file

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit loads tensor and optimizer data from the checkpoint file.

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Related issue: #13670
Draft: #13561

### CODE REVIEW

**Comment by hseok-oh** [comment]
Please resolve below 8 static analyzer warnings 

---

WGID 245566 | CheckpointLoader.cc:135 SIGN_EXTENSION
Unsafe conversion of expression '_header.other_offset' with type uint32_t to implementation defined type 'std::basic_istream::off_type' (aka 'std::streamoff' with 2 typedef(s) inbetween) may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

WGID 245765 | CheckpointLoader.cc:219 SIGN_EXTENSION
Unsafe conversion of expression '_optimizer_data[i][vindex].offset' with type uint32_t to implementation defined type 'std::basic_istream::off_type' (aka 'std::streamoff' with 2 typedef(s) inbetween) may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

WGID 245766 | CheckpointLoader.cc:163 SIGN_EXTENSION
Unsafe conversion of expression '_tensor_data[vindex].offset' with type uint32_t to implementation defined type 'std::basic_istream::off_type' (aka 'std::streamoff' with 2 typedef(s) inbetween) may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

WGID 245767 | CheckpointLoader.cc:125 SIGN_EXTENSION
Unsafe conversion of expression '_header.opt2_offset' with type uint32_t to implementation defined type 'std::basic_istream::off_type' (aka 'std::streamoff' with 2 typedef(s) inbetween) may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

WGID 245768 | CheckpointLoader.cc:164 SIGN_EXTENSION
Unsafe conversion of expression 'tensor->total_size()' with type size_t to implementation defined type 'std::streamsize' may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

WGID 245769 | CheckpointLoader.cc:126 SIGN_EXTENSION
Unsafe conversion of expression '_header.length * sizeof(uint32_t)' with type unsigned int to implementation defined type 'std::streamsize' may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

WGID 245770 | CheckpointLoader.cc:220 SIGN_EXTENSION
Unsafe conversion of expression 'opt_vars[i]->total_size()' with type size_t to implementation defined type 'std::streamsize' may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

WGID 245771 | CheckpointLoader.cc:116 SIGN_EXTENSION
Unsafe conversion of expression '_header.opt1_offset' with type uint32_t to implementation defined type 'std::basic_istream::off_type' (aka 'std::streamoff' with 2 typedef(s) inbetween) may or may not override sign-bit depending on size of implementation defined type, which may cause unexpected results on porting to different platforms
Major Classification : Undecided Component type: Default

File: runtime/onert/core/src/loader/train/CheckpointLoader.cc [hunk 1]
```diff
@@ -32,6 +32,48 @@ using namespace train;
...
+      _size[i] = _offset[i + 1] - _offset[i];
+    _size.back() = next_start_offset - _offset.back();
+  }
+
+  // offset, size
+  DataBufferPair operator[](uint32_t i) const
+  {
+    assert(_offset.size() == _size.size());
+    assert(i <= _offset.size());
+    return DataBufferPair{_offset[i], _size[i]};
```
Comment (ragmani) [inline runtime/onert/core/src/loader/train/CheckpointLoader.cc:69]: Unlike [std::map::operator[]](https://en.cppreference.com/w/cpp/container/map/operator_at), [std::vector::operator[]](https://en.cppreference.com/w/cpp/container/vector/operator_at) never inserts a new element into the container.
```suggestion
    assert(i < _offset.size());
    return DataBufferPair{_offset[i], _size[i]};
```

File: runtime/onert/core/src/loader/train/CheckpointLoader.cc [hunk 2]
```diff
@@ -76,10 +145,95 @@ class CheckpointLoader final
...
+private:
+  void updateAdamOptimizer(const std::unique_ptr<onert::exec::Execution> &exec)
+  {
+    // Adam optimizer has two optimizer variables. (mean, variance)
+    [[maybe_unused]] constexpr auto ADAM_VARIABLE_COUNT = 2;
+
+    // Reset EOF bit
+    _file.clear();
+
+    auto vindex = 0;
```
Comment (ragmani) [inline runtime/onert/core/src/loader/train/CheckpointLoader.cc:210]: (optional)
How about using an explicit type like the above `vindex`?
```suggestion
    uint32_t vindex = 0;
```

File: runtime/onert/core/src/loader/train/CheckpointLoader.cc [hunk 3]
```diff
@@ -76,10 +145,95 @@ class CheckpointLoader final
...
+  {
+    // TODO Verify cur_step value
+    train_info->trainingStep() = _footer.cur_step;
+  }
+
+private:
+  void updateAdamOptimizer(const std::unique_ptr<onert::exec::Execution> &exec)
+  {
+    // Adam optimizer has two optimizer variables. (mean, variance)
+    [[maybe_unused]] constexpr auto ADAM_VARIABLE_COUNT = 2;
```
Comment (ragmani) [inline runtime/onert/core/src/loader/train/CheckpointLoader.cc:205]: How about using an explicit type? This variable is used for comparing unsigned integer.

By the way, is this initializing variable is meaningful? this variable seems to be used in runtime, not compile time.

**Review by zetwhite — APPROVED** [review]
I carefully read this code. 
It looks good to me :+1:

**Comment by jyoungyun** [comment]
Sorry for late update. PTAL @ragmani @zetwhite  :)

**Review by ragmani — APPROVED** [review]
LGTM

**Review by zetwhite — APPROVED** [review]
LGTM

**Pull request merged**